src/macros/derivatives.js

```diff
@@ -0,0 +1,102 @@
+// @flow
+/**
+ * Macro utilities for assembling derivatives.
+ */
+
+import type {MacroContextInterface} from "../defineMacro";
+
+function splitByTopLevelComma(s: string): string[] {
+    if (!s || s.length === 0) {
+        return [];
+    }
+
+    // "1+a,n"         => ["1+a", "n"]
+    // "1+a,{n,m}"     => ["1+a", "{n,m}"]
+    // "1+a,{n,{p,q}}" => ["1+a", "{n,{p,q}}"]
+    // "1+a"           => ["1+a"]
+    // "1+a,,n"        => ["1+a", "", "n"]
+    // "1+a,,"         => ["1+a", ""]
+    // "1+a,"          => ["1+a"]
+    // ","             => [""]
+    // ""              => []
+    const res = [];
+    let currentStr = "";
+    let braceDepth = 0;
+    const isTopLevelDelimiter = (c) => c === "," && braceDepth <= 0;
+    for (let i = 0; i < s.length; ++i) {
+        const c = s[i];
+        if (c === "{") {
+            ++braceDepth;
+        } else if (c === "}") {
+            --braceDepth;
+        }
+        if (!isTopLevelDelimiter(c)) {
+            currentStr += c;
+        }
+        if (isTopLevelDelimiter(c) || i === s.length - 1) {
+            res.push(currentStr);
+            currentStr = "";
+        }
+    }
+
+    return res;
+}
+
+function addTokenProtector(s: string): string {
+    return s.length > 1 ? ` ${s} ` : s;
+}
+
+function extractArgs(context: MacroContextInterface): {
+    func: string,
+    vars: string[],
+    totalOrders: string[],
+    varOrders: string[],
+} {
+    const optionalArgGroups = [];
+    let arg = context.consumeArg().tokens;
+    while (arg.length === 1 && arg[0].text === "[") {
+        let bracketContent = '';
+        let token = context.expandNextToken();
+        while (token.text !== "]" && token.text !== "EOF") {
+            bracketContent += token.text;
+            token = context.expandNextToken();
+        }
+        optionalArgGroups.push(splitByTopLevelComma(bracketContent).reverse());
+        arg = context.consumeArg().tokens;
+    }
+
+    const func = arg.reverse().map(
+        e => addTokenProtector(e.text)).join("");
+    arg = context.consumeArg().tokens;
+    const vars = splitByTopLevelComma(arg.reverse().map(
+        e => addTokenProtector(e.text)).join(""));
+    const varOrders =
+        optionalArgGroups[0] ? optionalArgGroups[0].reverse() : ["1"];
+    const totalOrders =
+        optionalArgGroups[1] ? optionalArgGroups[1].reverse() : varOrders;
+
+    return {func, vars, totalOrders, varOrders};
+}
+
+const PLACEHOLDER = "\\square";
```

Comment:
Is this placeholder behavior implemented in any of the existing LaTeX packages? I didn't see it anywhere, but I may have missed it in one of them.

Reply:
No it's not implemented in any existing LaTeX package. That said, I think it's a useful improvement, so that users won't be puzzled by blankness.

Reply:
The square is a meaningful operator in many contexts and that might confuse more than it helps. I suppose if someone leaves the [options] blank they should expect that location in fact to be blank.